File: day10_Flow_network/python/part1.py
import re

class bot(object):
    def __init__(self, number):
        self.number = number
        self.chips = []
        self.high = None
        self.low = None

def chips_at_outputs(outputs):
    total = 0
    for key in outputs:
        total += len(outputs[key].chips)
    return total

def load_chips(filename, bots, outputs):
    chip_count = 0
    straightRGX = r"value (\d+) goes to bot (\d+)"
    conditionalRGX = r"bot (\d+) gives low to (output|bot) (\d+) and high to (output|bot) (\d+)"
    with open(filename) as f:
        active_bots = []
        for line in f:
            m = re.search(straightRGX, line)
            if m != None:
                bots[int(m.group(2))].chips.append(int(m.group(1)))
                chip_count += 1
            m = re.search(conditionalRGX, line)
            #using eval could be considered bad practice but
            if m != None:
                bots[int(m.group(1))].low = \
                    eval(m.group(2) + "s")[int(m.group(3))]
                bots[int(m.group(1))].high = \
                    eval(m.group(4) + "s")[int(m.group(5))]
    return bots, chip_count

def enable_flow(bots, outputs, chip_count):
    run = True
    from_61_to_17 = 0
    while chips_at_outputs(outputs) < chip_count:
        run = False
        del_keys = []
        for key in bots:
            if 61 in bots[key].chips and 17 in bots[key].chips:
                from_61_to_17 = bots[key].number
            if len(bots[key].chips) == 2:
                dest_high = bots[key].high
                dest_low = bots[key].low
                if bots[key].chips[0] - bots[key].chips[1] < 0 :
                    dest_high.chips.append(bots[key].chips[1])
                    dest_low.chips.append(bots[key].chips[0])
                else:
                    dest_low.chips.append(bots[key].chips[1])
                    dest_high.chips.append(bots[key].chips[0])
                del_keys.append(key)
        for key in del_keys:
            del bots[key]
    return outputs, from_61_to_17

def run(filename="input.txt"):
    bots = dict()
    outputs = dict()
    for i in range(220):
        bots[i] = bot(i)
    for i in range(22):
        outputs[i] = bot(i)
    bots, chip_count = load_chips(filename, bots, outputs)
    return enable_flow(bots, outputs, chip_count)

if __name__ == "__main__":
    outputs, from_61_to_17 = run()
    print(outputs[0].chips[0]*outputs[1].chips[0]*outputs[2].chips[0])
    print(from_61_to_17)

File: day10_Flow_network/cpp/part1.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <algorithm>
#include <vector>
#include <regex>
#include <chrono>
//Why can't you string switch!! lol

using namespace std;
using namespace std::chrono;
struct bot{
    int number=0;
    vector<int> chips;
    bot *low;
    bot *high;
};

int run(){
    ifstream txtFile("input.txt");
    if(txtFile.is_open()){
        int chip_count = 0;
        vector<bot> bots(220);
        for (int i=0; i < bots.size(); ++i){
            bots[i].number = i;
        }
        vector<bot> outputs(21);
        string line;
        smatch match;
        regex loadRGX("value (\\d+) goes to bot (\\d+)");
        regex flowRGX("(\\d+).*(bot|output) (\\d+).*(bot|output) (\\d+)");
        while(!txtFile.eof()){ //build network
                getline(txtFile, line);
                if (regex_search (line, match, loadRGX)){
                    bots[stoi(match[2])].chips.push_back(stoi(match[1]));
                    chip_count++;
                } else if (regex_search (line, match, flowRGX)){
                    bot *selected_bot = &bots[stoi(match[1])];
                    if (match[2].compare("bot") == 0){
                        selected_bot->low = &bots[stoi(match[3])];
                    } else {
                        selected_bot->low = &outputs[stoi(match[3])];
                    }
                    if (match[4].compare("bot") == 0){
                        selected_bot->high = &bots[stoi(match[5])];
                    } else {
                        selected_bot->high = &outputs[stoi(match[5])];
                    }
                }
            }
        txtFile.close();
        // let network propogate
        int output_count = 0;
        while (output_count < chip_count){
            //for some reason I couldn't get the iter to work?
            for (int b=0; b<bots.size(); ++b){
                bot *b_t = &bots[b];
                if (b_t->chips.size() == 2){
                    auto maxMin =
                        minmax_element(b_t->chips.begin(),b_t->chips.end());
                    if (*maxMin.first == 17 && *maxMin.second == 61){
                        cout << b_t->number << endl;
                    }
                    b_t->low->chips.push_back(*maxMin.first);
                    b_t->high->chips.push_back(*maxMin.second);
                    b_t->chips.push_back(0);
                }
            }
            output_count = 0;
            for (auto o:outputs){
                output_count += o.chips.size();
            }
        }
        int total = 1;

        for (int i = 0; i<outputs.size(); ++i){
            cout << i << ": " <<outputs[i].chips[0] << endl;
            if (i == 0 || i == 1 ||i == 2){
                total *= outputs[i].chips[0];
            }
        }
        cout << total << endl;

    }
}

int main()
{
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    run();
    high_resolution_clock::time_point t2 = high_resolution_clock::now();

    auto duration = duration_cast<milliseconds>( t2 - t1 ).count();

    cout << duration << " milliseconds" << endl;
    return 0;
}
